### DISCUSSION TITLE
Why does match-paths.json not include SSR page with matchPath?

### DISCUSSION THREAD

**Discussion by supachailllpay** [gatsbyjs/gatsby#35864] [Help]
For example, I have a page `/src/pages/foo/bar/[id].js` (SSR).

And the page object will look something like this:

```js
{
  path: '/foo/bar/[id]/',
  matchPath: '/foo/bar/:id'
}
```

Next, I would like to have a custom 404 page for any path that doesn't exist for /foo/*.

So I created another page `/src/pages/foo/404.js` (SSG):

```js
{
  path: '/foo/404/',
  matchPath: undefined
}
```

https://github.com/gatsbyjs/gatsby/blob/8257bca2a1d352fef7d4a261926e7d455154bf2e/packages/gatsby/src/bootstrap/requires-writer.ts#L109-L121

From the code above, the `match-paths.json` still be an empty array because the `[id].js` is SSR not SSG and `404.js` doesn't have a matchPath.

Everything works fine except the 404 page doesn't match other paths yet. So I added the code in `onCreatePage` to modify the matchPath of `404.js` and now the page object looks like this:

```js
{
  path: '/foo/404/',
  matchPath: '/foo/*'
}
```

Since `404.js` now has a matchPath so the `match-paths.json` is:

```json
[
  {
    "path": "/foo/404/",
    "matchPath": "/foo/*"
  }
]
```

And now `[id].js` is no longer accessible, it always goes to `404.js` instead. But if I change `[id].js` to SSG, everything is back to working normally (`match-paths.json` contains `/foo/bar/[id]/` path).

Or if I have other SSR pages that don't have matchPath, they will be included in the `match-paths.json` and it works fine.

So my question is, why is an SSR page with matchPath (client-only routes) not included in the `match-paths.json` and how to solve this?

**Comment by danilobuerger** [comment]
At least for me it still seems to work when using `serve`, but not `develop`

**Comment by rwitchell** [comment]
`[id].js` isn't SSR. it's client-side only, so gatsby ignores this page when your run `gatsby build`. 

instead, if you want to *actually* generate SSR: you should look at `{product.id}.js` or using gastby-node 

```js
result.products.forEach(product => {
      createPage({
        path: `thing/${product.id}/`,
        component: path.resolve('src/templates/product.js'),
        context: {
          id: product.id,
          product,
        },
        defer: true,
      })
    })
```

**Reply by supachailllpay** [reply]
`[id].js` is a client-only route, according to the [documentation](https://www.gatsbyjs.com/docs/reference/routing/file-system-route-api/#syntax-client-only-routes) and it's also an SSR page as the result of build:

```
├ src/pages/campaigns/[slug].js
│ ├ ∞ /en/campaigns/[slug]/
│ └ ∞ /campaigns/[slug]/

  ╭────────────────────────────────────────────────────────────────╮
  │                                                                │
  │   (SSG) Generated at build time                                │
  │ D (DSG) Deferred static generation - page generated at runtime │
  │ ∞ (SSR) Server-side renders at runtime (uses getServerData)    │
  │ λ (Function) Gatsby function                                   │
  │                                                                │
  ╰────────────────────────────────────────────────────────────────╯
```
However, my point is why only the page with `getPageMode(page) === 'SSG'` is included in `matchPathPages` why not just any page with `matchPath`?
